web/modules/custom/joinup_core/src/EventSubscriber/RefreshCachedFieldsEventSubscriber.php

```diff
@@ -86,89 +86,81 @@ public function __construct(EntityTypeManagerInterface $entityTypeManager, TimeI
+      $type = $this->getType($bundle);
+      $response_item = $response[$index];
+      $count = 0;
+      foreach ($response_item as $year => $result) {
+        $item = $result[0];
+        if (!empty($item->$type)) {
+          $count = $count + (int) $item->$type;
+        }
+      }
+
+      $this->updateFieldValue($expired_item, $count);
     }
   }
 
   /**
-   * Returns the action count the given entity received in the given period.
+   * Gets the correct URL parameter for the query.
    *
    * @param \Drupal\Core\Entity\ContentEntityInterface $entity
-   *   The entity to check.
+   *   The entity from which to prune the URL.
    *
-   * @return int|false
-   *   The number of visits, or FALSE if the number of visits could not be
-   *   determined.
+   * @return string
+   *   The url for the request.
+   *
+   * @throws \Drupal\joinup_core\Exception\JoinupCoreUndefinedUrlException
+   *   Thrown when a URL could not be distilled from the given entity.
    */
-  protected function getCount(ContentEntityInterface $entity) {
+  protected function getUrlParameter(ContentEntityInterface $entity) {
     $bundle = $entity->bundle();
-    $period = $this->getTimePeriod($bundle);
-    $type = $this->getType($bundle);
-    $method = $this->getPiwikMethod($bundle);
-
-    // Start building the query.
-    $query = $this->piwikQueryFactory->getQuery($method);
+    switch ($this->getMethod($bundle)) {
+      case 'download_counts':
+        $url = $this->getFileUrl($entity);
+        break;
 
-    // Try to retrieve the URL from the entity.
-    try {
-      $this->setUrlParameter($query, $entity);
-    }
-    catch (JoinupCoreUndefinedUrlException $e) {
-      // If no URL was found then we cannot get results from Piwik. This can
-      // happen for example when retrieving the download count from a
-      // distribution that has no files associated with it.
-      // We are returning 0 instead of FALSE so that this field will not be
-      // re-queued on every cron run for eternity.
-      return 0;
+      default:
+        $url = $this->getEntityUrl($entity);
+        break;
     }
 
-    $date_range = [
-      // If the period is 0 we should get all results since launch.
-      $period > 0 ? (new DateTimePlus("$period days ago"))->format('Y-m-d') : $this->configFactory->get('joinup_core.piwik_settings')->get('launch_date'),
-      (new DateTimePlus())->format('Y-m-d'),
-    ];
-
-    $query->setParameters([
-      'period' => 'year',
-      'date' => implode(',', $date_range),
-      'showColumns' => $type,
-    ]);
-    $response = $query->execute();
-    if (!$response->hasError()) {
-      $response = $response->getResponse();
-      $count = 0;
-      foreach ($response as $year => $result) {
-        $item = $result[0];
-        if (!empty($item->$type)) {
-          $count = $count + (int) $item->$type;
-        }
-      }
-      return $count;
+    if (empty($url)) {
+      throw new JoinupCoreUndefinedUrlException();
```

Comment:
Please add a comment explaining when this can occur.

Reply:
I will remove this entirely since it is meant to catch distributions without a file mainly because in any other case, the url will be present.
I will have the entity updated already with a count of 0 if that case is reached in order to have it excluded from the next update effort.